### PR TITLE
feat: chain watcher

### LINKED ISSUE #46: feat: implement own event watcher / block watcher

### PR DESCRIPTION
**Pull request by mfw78**
# Description
As we are removing Tenderly, we need to handle our own run loops for watching blocks / transaction logs.

# Changes

- [x] Add `warmUp` to initialise state from genesis, or last processed block.
- [x] Used rebuild logic from #29 for `warmUp`
- [x] Once `warmUp` is complete, flag as `inSync`, and commence block watching.
- [x] Implemented all in `ChainWatcher`, which runs per-chain, but we can handle many RPCs
- [x] All chains watched from the same instance.

## How to test

No unit tests available, however can observe functionality, such as with:

```bash
yarn ts-node ./src/watchtower.ts run --rpc http://erigon.dappnode:8545 --rpc http://nethermind-xdai.dappnode:8545 --deployment-block 17883049 --deployment-block 29389123 --page-size 0
```

**NOTE**: Substitute the above RPCs with the RPCs of your choice for the chains.

Observe synchronising of state. If Ctrl+C, observe `SIGINT` is caught, database is saved. Restart, and node resumes from last known synced block.

## Todo

- [x] Check for TODOs / dead code.

## Related Issues

Closes #46

### CODE REVIEW

File: src/modes/run.ts [hunk 1]
```diff
@@ -1,5 +1,340 @@
...
+
+    // Run all the chain watchers
+    await Promise.all(runPromises);
+  } catch (error) {
+    console.error(error);
+    await DBService.getInstance().close();
+    process.exit(1);
+  } finally {
+    await DBService.getInstance().close();
+  }
```
Comment (alfetopito) [inline src/modes/run.ts:56]: If I'm not mistaken, finally will only run in a regular execution, since catch exits before it, no?
Either way, the DB is already closed in that case.
Might just move line `55` after `49` and remove the finally

File: src/modes/run.ts [hunk 2]
```diff
@@ -1,5 +1,340 @@
...
+}
+
+/**
+ * The chain watcher handles watching a single chain for new conditional orders
+ * and executing them.
+ */
+export class ChainWatcher {
```
Comment (alfetopito) [inline src/modes/run.ts:63]: This should live in its own file IMO

File: src/modes/run.ts [hunk 3]
```diff
@@ -1,5 +1,340 @@
...
+  /**
+   * Warm up the chain watcher by fetching the latest block number and
+   * checking if the chain is in sync.
+   * @returns the run promises for what needs to be watched
+   */
+  public async warmUp(oneShot?: boolean) {
+    const { provider, chainId } = this.chainContext;
+    const { lastProcessedBlock } = this.registry;
+    const { pageSize } = this;
+    const _ = (s: string) => console.log(`[warmUp:chainId:${chainId}] ${s}`);
```
Comment (alfetopito) [inline src/modes/run.ts:111]: Please no variables named `_` 😆 

If you intend to keep it, make it a private method that returns the log fn for example.

```ts
private getLog(chainId:number): (s:string)=>void {
  return (s:string) => console.log(`[warmUp:chainId:${chainId}] ${s}`)
}
```

File: src/modes/run.ts [hunk 4]
```diff
@@ -1,5 +1,340 @@
...
+    let currentBlockNumber = await provider.getBlockNumber();
+
+    let plan: ReplayPlan = {};
+    let toBlock: "latest" | number = 0;
+    do {
+      do {
+        toBlock = !pageSize ? "latest" : fromBlock + (pageSize - 1);
+        if (typeof toBlock === "number" && toBlock > currentBlockNumber) {
+          // refresh the current block number
+          currentBlockNumber = await provider.getBlockNumber();
```
Comment (alfetopito) [inline src/modes/run.ts:126]: Would it make sense to refactor a bit some of this code into helper fns/methods?

A bit hard to follow for example where the while loops end

File: src/modes/run.ts [hunk 5]
```diff
@@ -1,5 +1,340 @@
...
+        fromBlock = this.registry.lastProcessedBlock + 1;
+        plan = {};
+      }
+    } while (!this.inSync);
+
+    _(oneShot ? "Chain watcher is in sync" : "Chain watcher is warmed up");
+    _(`Last processed block: ${this.registry.lastProcessedBlock}`);
+
+    // Notifying that the chain watcher is in sync
+    this.inSync = true;
```
Comment (alfetopito) [inline src/modes/run.ts:209]: Again?
This is done as the exit condition for the loop on line 196, thus already set by this point.

File: src/modes/run.ts [hunk 6]
```diff
@@ -1,5 +1,340 @@
...
+   * Run the block watcher for the chain. As new blocks come in:
+   * 1. Check if there are any `ConditionalOrderCreated` events, and index these.
+   * 2. Check if any orders want to create discrete orders.
+   */
+  public async runBlockWatcher() {
+    const { provider, chainId } = this.chainContext;
+    const _ = (s: string) =>
+      console.log(`[runBlockWatcher:chainId:${chainId}] ${s}`);
+    // Watch for new blocks
+    _("Subscribe to new blocks");
```
Comment (alfetopito) [inline src/modes/run.ts:230]: 👀

File: src/modes/run.ts [hunk 7]
```diff
@@ -1,5 +1,340 @@
...
+        .catch((e) => {
+          hasErrors = true;
+          console.error(
+            `[run_local] Error running "addContract" action for TX:`,
+            e
+          );
+          return false;
+        });
+      console.log(
+        `[run_local] Result of "addContract" action for TX ${
```
Comment (alfetopito) [inline src/modes/run.ts:289]: There 2 logs do not use the same prefix as the other log msgs.
Is this intentional?

**Review by anxolin — APPROVED** [review]
Worked great

File: src/addContract.ts [hunk 8]
```diff
@@ -1,69 +1,63 @@
...
- * @param context tenderly context
+ * @param chainWatcher chain watcher
  * @param event transaction event
  */
-export const addContract: ActionFn = async (context: Context, event: Event) => {
-  return _addContract(context, event).catch(handleExecutionError);
+export const addContract = async (
+  chainWatcher: ChainWatcher,
+  event: ConditionalOrderCreatedEvent
+) => {
```
[the remark this answers is not in the document]
Comment (mfw78) [reply src/addContract.ts:30]: Done.

I'm a fan of old school things like "function" :sweat_smile:

File: src/modes/run.ts [hunk 9]
```diff
@@ -1,5 +1,340 @@
...
+
+/**
+ * The chain watcher handles watching a single chain for new conditional orders
+ * and executing them.
+ */
+export class ChainWatcher {
+  private readonly deploymentBlock: number;
+  private readonly pageSize: number;
+  private readonly publish: boolean;
+  private inSync = false;
```
[the remark this answers is not in the document]
Comment (mfw78) [reply src/modes/run.ts:67]: Tracked in #60

File: src/modes/run.ts (hunk 2 above)
Comment (mfw78) [reply src/modes/run.ts:63]: Tracked in #60

File: src/modes/run.ts (hunk 4 above)
Comment (mfw78) [reply src/modes/run.ts:126]: Tracked in #60

File: src/modes/run.ts (hunk 7 above)
Comment (mfw78) [reply src/modes/run.ts:289]: Tracked in #60

File: src/modes/run.ts (hunk 6 above)
Comment (mfw78) [reply src/modes/run.ts:230]: Heh, your inquisitiveness is tracked in #60

File: src/modes/run.ts (hunk 3 above)
Comment (mfw78) [reply src/modes/run.ts:111]: Unified logging tracked in #60

File: src/modes/run.ts (hunk 1 above)
Comment (mfw78) [reply src/modes/run.ts:56]: Fixed.

File: src/modes/run.ts [hunk 10]
```diff
@@ -1,5 +1,340 @@
...
-  throw new Error("Not implemented");
+/**
+ * Run the watch-tower 👀🐮
+ * @param options Specified by the CLI / environment for running the watch-tower
+ */
+export async function run(options: RunOptions) {
+  const { rpc, deploymentBlock, oneShot } = options;
+
+  process.on("unhandledRejection", async (error) => {
+    console.log(error);
```
[the remark this answers is not in the document]
Comment (mfw78) [reply src/modes/run.ts:18]: Logging to be reviewed in #60

File: src/types/model.ts [hunk 11]
```diff
@@ -253,10 +254,7 @@ export class ChainContext {
     this.chainId = chainId;
   }
 
-  public static async create(
-    storage: DBService,
```
[the remark this answers is not in the document]
Comment (mfw78) [reply src/types/model.ts:253]: I think this is a little strange as I show:

```typescript
  public static async create(url: string): Promise<ChainContext> {
    const provider = new ethers.providers.JsonRpcProvider(url);
    const chainId = (await provider.getNetwork()).chainId;
    return new ChainContext(provider, apiUrl(chainId), chainId);
  }
```

Around this function. Nonetheless, can be added in a later PR to get over the waterfall :water_buffalo:
